formulaic/transforms/basis_spline.py

```diff
@@ -122,10 +124,22 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
             "Some field values extend beyond upper and/or lower bounds, which can result in ill-conditioned bases. "
             "Pass a value for `extrapolation` to control how extrapolation should be performed."
         )
-    if extrapolation is SplineExtrapolation.CLIP:
-        x = numpy.clip(x, lower_bound, upper_bound)
-    if extrapolation is SplineExtrapolation.NA:
-        x = numpy.where((x >= lower_bound) & (x <= upper_bound), x, numpy.nan)
+    if extrapolation in (
+        SplineExtrapolation.CLIP,
+        SplineExtrapolation.NA,
+        SplineExtrapolation.ZERO,
+    ):
+        locs = (x >= lower_bound) & (x <= upper_bound)
```

Comment:
This avoids copies in most applications IME, but added some complexity to the code.